src/netbase.h

```diff
@@ -37,6 +41,101 @@ class proxyType
     bool randomize_credentials;
 };
 
+/**
+ * Convert milliseconds to a struct timeval for e.g. select.
+ */
+struct timeval MillisToTimeval(int64_t nTimeout);
+
+// Represents a read/write socket.
+class MockableSocket
+{
+public:
+    virtual ssize_t send(const char* buf, size_t len, int flags) = 0;
+
+    virtual ssize_t recv(char* buf, size_t len, int flags) = 0;
+
+#ifdef USE_POLL
+    virtual int poll(short events, int timeout_ms) = 0;
+#endif
+
+    virtual int select(bool watch_read, bool watch_write, int timeout_ms) = 0;
+
+    virtual bool IsSelectable() const = 0;
+};
+
+// Represents a non-owned SOCKET.
+//
+// The underlying SOCKET will NOT be closed automatically on object destruction.
+class Socket : public MockableSocket
+{
+    SOCKET m_socket = INVALID_SOCKET;
+
+public:
+    explicit Socket(SOCKET socket) : m_socket{socket}
+    {
+    }
+
+    // Send bytes on the socket. See send(2).
+    ssize_t send(const char* buf, size_t len, int flags) override
+    {
+        return ::send(m_socket, buf, len, flags);
+    }
+
+    // Receive bytes from the socket. See recv(2).
+    ssize_t recv(char* buf, size_t len, int flags) override
+    {
+        return ::recv(m_socket, buf, len, flags);
+    }
+
+#ifdef USE_POLL
```

Comment:
I'd prefer to have this implementation in an implementation unit instead of the header. This is not a templated class and overriding virtual methods can be done everywhere.
This would also avoid adding the optional include `poll.h` here, I think.

Reply:
Good point! Pushed an updated version addressing this.